src/go/rpk/pkg/cli/topic/analyze.go

```diff
@@ -0,0 +1,743 @@
+}
+
+func (c *collector) results() collectResult {
+	return collectResult{c.tpInfo, c.cfg.TimeRange}
+}
+
+type percentileValue[V any] struct {
+	percentile float64
+	value      V
+}
+
+type topicSummary struct {
+	Topic            string
+	Partitions       int
+	BatchesPerSecond []percentileValue[float64]
+	BatchSizes       []percentileValue[int] // bytes
+}
+
+func (cr *collectResult) summarizeTopics(percentiles []float64) ([]topicSummary, error) {
+	totalDuration := cr.TimeRange.End.Sub(cr.TimeRange.Start)
+	tps := map[string]struct {
+		Partitions []int
+		BatchSizes []int
+		BatchRates []float64
+	}{}
+	for tp, info := range cr.TopicPartitionInfo {
+		e := tps[tp.Topic]
+		e.Partitions = append(e.Partitions, int(tp.Partition))
+
+		if info.empty() {
+			tps[tp.Topic] = e
+			continue
+		}
+
+		batchesPerS := float64(info.BatchesRead) / totalDuration.Seconds()
+		e.BatchRates = append(e.BatchRates, batchesPerS)
+
+		if !slices.IsSorted(info.BatchSizes) {
+			slices.Sort(info.BatchSizes)
+		}
+
+		p50, err := utils.Percentile(info.BatchSizes, 50.0)
+		if err != nil {
+			return nil, err
+		}
+		// Note that we insert the p50 batch size for each partition here.
+		// Another option could be inserting every sampled batch size then
+		// take percentiles of that. However, since each partition may have
+		// a different number of sampled batches care would need to be taken
+		// to insert the same number of samples per partition to avoid bias.
+		e.BatchSizes = append(e.BatchSizes, p50)
+
+		tps[tp.Topic] = e
+	}
+
+	tss := []topicSummary{}
+	for t, i := range tps {
+		ts := topicSummary{Topic: t, Partitions: len(i.Partitions)}
+
+		slices.Sort(i.BatchSizes)
+		slices.Sort(i.BatchRates)
+		for _, p := range percentiles {
+			if len(i.BatchSizes) != 0 {
+				pbs, err := utils.Percentile(i.BatchSizes, p)
+				if err != nil {
+					return nil, err
+				}
+				ts.BatchSizes = append(ts.BatchSizes, percentileValue[int]{p, pbs})
+			} else {
+				ts.BatchSizes = append(ts.BatchSizes, percentileValue[int]{p, 0})
+			}
+
+			if len(i.BatchRates) != 0 {
+				pbr, err := utils.Percentile(i.BatchRates, p)
+				if err != nil {
+					return nil, err
+				}
+				ts.BatchesPerSecond = append(ts.BatchesPerSecond, percentileValue[float64]{p, pbr})
+			} else {
+				ts.BatchesPerSecond = append(ts.BatchesPerSecond, percentileValue[float64]{p, 0})
+			}
+		}
+
+		tss = append(tss, ts)
+	}
+
+	return tss, nil
+}
+
+type globalSummary struct {
+	Topics                int
+	Partitions            int
+	TotalBatchesPerSecond float64
+	BatchesPerSecond      []percentileValue[float64]
+	BatchSizes            []percentileValue[int] // bytes
+}
+
+func (cr *collectResult) summarizeGlobal(percentiles []float64) (globalSummary, error) {
+	summary := globalSummary{}
+	parts := 0
+	totalBatches := 0
+	totalDuration := cr.TimeRange.End.Sub(cr.TimeRange.Start)
+	topics := map[string]struct{}{}
+	batchSizes := []int{}
+	batchRates := []float64{}
```

Comment:
Here is fine to leave the initialization explicitly since it will be used to be printed by the json unmarshaller, which will print `[]` here instead of `nil` if you used `var batchSizes []int` :+1: